File: core/src/ru/myitschool/dcrawler/utils/AdvancedArray.java
package ru.myitschool.dcrawler.utils;

import com.badlogic.gdx.utils.Array;

import java.util.Random;

/**
 * Array with additional functions of clipping, getting random element and stack/queue-style getters
 */
public class AdvancedArray<T> extends Array<T> {

    public boolean clip(int from, int to){
        if (from >= 0 && to < items.length && from <= to) {
            T[] objects = (T[]) new Object[to - from + 1];
            for (int i = 0; i < to - from + 1; i++) {
                objects[i] = items[from + i];
            }
            items = objects;
            size = objects.length;
            return true;
        }
        return false;
    }

    public T getFirst(){
        return items[0];
    }

    public T getLast(){
        return items[items.length - 1];
    }

    public T getRandom() {
        int random = SeededRandom.getInstance().nextInt(size);
        return items[random];
    }
}

File: core/src/com/wolg_vlad/dcrawler/utils/AdvancedArray.kt
package com.wolg_vlad.dcrawler.utils

/**
 * Array with additional functions of clipping, getting random element and stack/queue-style getters
 */
class AdvancedArray<T> : com.badlogic.gdx.utils.Array<T>() {
    fun clip(from: Int, to: Int): Boolean {
        if (from >= 0 && to < items.size && from <= to) {
            val objects = arrayOfNulls<Any>(to - from + 1) as Array<T?>
            for (i in 0 until to - from + 1) {
                objects[i] = items[from + i]
            }
            items = objects
            size = objects.size
            return true
        }
        return false
    }

    val first: T
        get() = items[0]

    val last: T
        get() = items[size - 1]

    val random: T
        get() {
            val random = SeededRandom.getInstance()!!.nextInt(size)
            return items[random]
        }
}